mx-find-linked-structures Assertion Error

Running `mx-find-linked-structures` on an indexed Linux kernel, and running into an assertion error:

```
multiplier/lib/SQLiteEntityProvider.cpp:1082: virtual DeclImplPtr mx::SQLiteEntityProvider::DeclFor(const Ptr &, RawEntityId): Assertion `false' failed.

Thread 1 "mx-find-linked-" received signal SIGABRT, Aborted.
```

Built a debug version and it looks like it starts from [FindLinkedStructures.cpp:76](https://github.com/trailofbits/multiplier/blob/90a3c819e9c66dd42cbc28b6df35b839df029af4/bin/Examples/FindLinkedStructures.cpp#L76) in the first phase of analysis. Happy to provide more debugging info if that would be helpful.

Backtrace:
```
#0  __pthread_kill_implementation (no_tid=0x0, signo=0x6, threadid=0x7ffff7e4e100)
    at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=0x6, threadid=0x7ffff7e4e100) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=0x7ffff7e4e100, signo=signo@entry=0x6)
    at ./nptl/pthread_kill.c:89
#3  0x00007fffe6042476 in __GI_raise (sig=sig@entry=0x6) at ../sysdeps/posix/raise.c:26
#4  0x00007fffe60287f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007fffe602871b in __assert_fail_base (
    fmt=0x7fffe61dd130 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=0x7fffe90bb20b "false",
    file=0x7fffe90bdcaf "/home/user0/git/mx-work/src/multiplier/lib/SQLiteEntityProvider.cpp", line=0x43a, function=<optimized out>) at ./assert/assert.c:92
#6  0x00007fffe6039e96 in __GI___assert_fail (assertion=0x7fffe90bb20b "false",
    file=0x7fffe90bdcaf "/home/user0/git/mx-work/src/multiplier/lib/SQLiteEntityProvider.cpp", line=0x43a,
    function=0x7fffe98353b4 "virtual DeclImplPtr mx::SQLiteEntityProvider::DeclFor(const Ptr &, RawEntityId)") at ./assert/assert.c:101
#7  0x00007ffff00165cd in mx::SQLiteEntityProvider::DeclFor (this=0x5555557bb9a0,
    self=std::shared_ptr<mx::EntityProvider> (use count 97, weak count 1) = {...},
    raw_id=0x84a000078e000001) at multiplier/lib/SQLiteEntityProvider.cpp:1075
#8  0x00007fffefd9f595 in mx::CachingEntityProvider::DeclFor (this=0x5555557e7cb0,
    self=std::shared_ptr<mx::EntityProvider> (use count 97, weak count 1) = {...},
    raw_id=0x84a000078e000001) at multiplier/lib/CachingEntityProvider.cpp:279
#9  0x00007fffef8af1c9 in mx::TagType::declaration (this=0x7fffffffde38)
    at multiplier/lib/AST/TagType.cpp:107
#10 0x0000555555592bfc in main (argc=0x3, argv=0x7fffffffe228)
    at multiplier/bin/Examples/FindLinkedStructures.cpp:76
```


## System info
Ubuntu 22.04.4
multiplier tag: `c18052b`

Currently struggling to complete a clean build at the moment. Keep getting a bunch of undefined reference errors when linking:

```
ld.lld: error: lib/libmultiplier.so: undefined reference to llvm::EnableABIBreakingChecks [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to llvm::ilist_detail::SpecificNodeAccess<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false> >::getNodePtr(mlir::Operation*) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to llvm::ilist_detail::SpecificNodeAccess<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false> >::getValuePtr(llvm::ilist_node_impl<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false> >*) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::RewriterBase::inlineRegionBefore(mlir::Region&, mlir::Region&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::RewriterBase::cloneRegionBefore(mlir::Region&, mlir::Region&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>, mlir::IRMapping&) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::RewriterBase::inlineBlockBefore(mlir::Block*, mlir::Block*, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>, mlir::ValueRange) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::RewriterBase::splitBlock(mlir::Block*, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to llvm::ilist_traits<mlir::Block>::transferNodesFromList(llvm::ilist_traits<mlir::Block>&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to llvm::ilist_traits<mlir::Operation>::transferNodesFromList(llvm::ilist_traits<mlir::Operation>&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::ConversionPatternRewriter::inlineRegionBefore(mlir::Region&, mlir::Region&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::OpBuilder::createBlock(mlir::Region*, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>, mlir::TypeRange, llvm::ArrayRef<mlir::Location>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::RewriterBase::cloneRegionBefore(mlir::Region&, mlir::Region&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::ConversionPatternRewriter::splitBlock(mlir::Block*, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
ld.lld: error: lib/libmultiplier.so: undefined reference to mlir::Block::splitBlock(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void, false>, false, false>) [--no-allow-shlib-undefined]
clang++-18: error: linker command failed with exit code 1 (use -v to see invocation)
[1580/3736] Building CXX object bin/Examples/C...iles/mx-list-functions.dir/ListFunctions.cpp.o
ninja: build stopped: subcommand failed.
```
This is coming after blowing away everything in build, install, and src and starting from scratch from the instructions in `BUILD.md`. Not sure exactly what's going on here.

--- a/bin/Examples/CompressCompilation.py
+++ b/bin/Examples/CompressCompilation.py
@@ -41,16 +41,14 @@ def harness(comp: mx.frontend.Compilation, root_path: pathlib.PurePath):
             "resource_directory": str(comp.resource_directory),
         }], indent="  "))
 
-    # Go get all macro definitions not associated with a file. This will capture
-    # compiler-builtin macros. Unfortunately this will also capture the macros
-    # defined at the command-line.
+    # Go get all (used) compiler builtin macro definitions.
     with open(str(root_path / "builtin-macros.h"), "w") as macros:
-        index = mx.Index.containing(comp)
-        for macro in mx.frontend.DefineMacroDirective.IN(index):
-            file = mx.frontend.File.containing(macro)
-            if not file:
-                macros.write(" ".join(t.data for t in macro.generate_use_tokens))
-                macros.write('\n')
+        for macro in comp.builtin_defines:
+            macros.write(f"#ifdef {macro.name.data}\n")
+            macros.write(f"#  undef {macro.name.data}\n")
+            macros.write("#endif\n")
+            macros.write(" ".join(t.data for t in macro.generate_use_tokens))
+            macros.write('\n')
 
     with tarfile.open(str(root_path) + ".tar.gz", "w:gz") as tar:
         tar.add(str(root_path), arcname=root_path.name)

